---
title: "GO enrichment sheet"
output: word_document
date: "2025-03-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,        # Show code (set to FALSE to hide)
  message = FALSE,    # Hide package loading messages
  warning = FALSE,    # Hide warnings
  error = TRUE        # allow knitting to continue if errors occur
)
# setwd("J:/Dropbox/Prague/Teaching/Course genomics of speciation/2024-2025/RNA-Seq practical/GO enrichment")
```

Before starting, a few questions for you: what does a GO enrichment do? What can it tell us? By definition, an enrichment means an over-representation in a group compared to another? What are those two groups? These questions are important to answer so that you use the script below properly.

# Getting started - special installation

```{r}

#install clusterProfiler (special way; for the other packages, the normal way should work)
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("clusterProfiler")

#load packages, set up the working directory
library(clusterProfiler)
library(ggplot2)
library(tidyr)
library(dplyr)
getwd()
setwd("J:/Dropbox/Prague/Teaching/Course genomics of speciation/2024-2025/RNA-Seq practical/GO enrichment")
```

# Importing the reference GO files

```{r}
#The files should look like this: 
#GO:0055114	Solyc02g080590	Biological Process:oxidation-reduction process
#GO:0055114	Solyc08g007400	Biological Process:oxidation-reduction process
#GO:0055114	Solyc03g031650	Biological Process:oxidation-reduction process

tryCatch({
  SlycBP <- read.csv("SlycBP.csv")
  term2gene_BP <- SlycBP[,c(1,2)]
  term2name_BP <- SlycBP[,c(1,3)]
}, error = function(e) message("Problem loading SlycBP.csv"))
term2gene_BP <- SlycBP[,c(1,2)] #extract first 2 columns (GO term and gene ID)
term2name_BP <- SlycBP[,c(1,3)] #extract Go term and function

tryCatch({
  SlycCC <- read.csv("SlycCC.csv")
}, error = function(e) message("Problem loading SlycCC.csv"))
term2gene_CC <- SlycCC[,c(1,2)] #extract first 2 columns (GO term and gene ID)
term2name_CC <- SlycCC[,c(1,3)] #extract Go term and function

SlycMF <- read.csv("SlycMF.csv")
term2gene_MF <- SlycMF[,c(1,2)] #extract first 2 columns (GO term and gene ID)
term2name_MF <- SlycMF[,c(1,3)] #extract Go term and function

```

# Reading your list of genes

```{r}
#the file should look like this (only 1 column with gene names):
#Solyc03g031650
#Solyc03g031650
#Solyc03g031650
#Solyc03g031650

candidate_genes <- read.csv("XXXXX.csv")
candidate_genes = as.vector(candidate_genes[,1])


#Check if they intersect
intersect(term2gene_BP$locusName, candidate_genes)
```

# Testing for enrichment (PS: between what and what?)

```{r}
#BP
candidate_genes_BP <- enricher(gene= candidate_genes, pvalueCutoff = 0.05, pAdjustMethod = "fdr", TERM2GENE= term2gene_BP, TERM2NAME =  term2name_BP)
head(candidate_genes_BP)
candidate_genes_BP <- candidate_genes_BP[,c(2,3,4,6,8,9)] #function, p-value, geneIDs, count
candidate_genes_BP$GeneRatio <- paste0("'", candidate_genes_BP$GeneRatio) # Add an apostrophe before each GeneRatio value, otherwise it will be read as a date from excel
write.csv(candidate_genes_BP, 'candidate_genes_GO_BP.csv')

```

This was for Biological Processes. Now you're gonna need to do the same for Celllular Component and Molecular Function.

# Preparing the file with all enriched terms for plotting

This puts together the enriched GO terms in a csv file that looks like this:

Category Description GeneRatio p.adjust\
upregulated Molecular Function:DNA-binding transcription factor activity 1 0.023282605

```{r}
#Give a name to your group of candidate genes (as an example here is "upregulated"), add it as column to each of the enriched dfs; if you have several groups of genes, do that for every group
candidate_genes_BP$Category <- "upregulated"
candidate_genes_CC$Category <- "upregulated"
candidate_genes_MF$Category <- "upregulated"


All_GO <- rbind(candidate_genes_BP, 
                candidate_genes_CC,candidate_genes_MF) #if you have several groups of genes, you can put them all together here if you want them plotted together

```

# Plotting to visualize the enriched GO terms

This will generate dotplots showing gene ratio and adj pvalue.

```{r}

GO <- All_GO
GO$Description <- as.factor(GO$Description)

# Edit our GeneRatio column
# Remove the leading ' from the strings
GO$GeneRatio <- gsub("^'", "", GO$GeneRatio)
# Use separate() to split the column into two new columns

GO <- GO %>% separate(GeneRatio, into = c("Numerator", "Denominator"), sep = "/")
# Convert the new columns to numeric
GO$Numerator <- as.numeric(GO$Numerator)
GO$Denominator <- as.numeric(GO$Denominator)
GO$GeneRatio <- GO$Numerator/GO$Denominator
# Remove extra columns
GO$Numerator <- NULL
GO$Denominator <- NULL

#plot all
ggplot(data = GO, aes(x = Category, y = Description, 
                      color = `p.adjust`, size = GeneRatio)) + 
  geom_point() +
  scale_color_gradient(low = "red", high = "blue") +
  theme_bw() +
  ylab("") + 
  xlab("") +
  theme_bw() +
  #theme(axis.title.x = element_text(size = 24), axis.text.x = element_text(size = 24)) +
  theme(axis.text.y = element_text(size = 12)) + 
  #theme(legend.title = element_blank(), legend.text = element_text(size = 24)) + 
  #theme(axis.text.x = element_text(angle = 25, vjust = 0.8, hjust =1, size = 12)) +
  theme(axis.text.x = element_text(vjust = 0.8, hjust =1, size = 12)) +
  scale_y_discrete(limits=rev)
dev.off() #this closes the graph, so if you want to see the graph, don't run that :)
```
